### DISCUSSION TITLE
[Desktop] How to create multi window app that uses dedicated threads for each window?

### DISCUSSION THREAD

**Discussion by kucint** [unoplatform/uno#16899] [Q&A] [answered]
This document explains how to create multi window application: [Exploring multi-window support for Linux, macOS, and Windows](https://platform.uno/blog/exploring-multi-window-support-for-linux-macos-and-windows/)

I would like to run each of these windows is dedicated threads, one thread per window.
How am I supposed to do it?

**Comment by jeromelaban** [comment]
At this time, this is not supported. All window instance share the same dispatcher. @MartinZikmund will have more details.

**Reply by kucint** [reply]
that is a bad news :(.
Any chance to improve it?

**Reply by MartinZikmund** [reply]
The problem here is that WinUI itself does not have proper support for this scenario and as a result it is very complex to add in our case as well. Is there a specific reason why you require this? Non-UI operations can still be offloaded on a non-UI thread, so only the UI-bound ones will need to run on the UI thread

**Reply by kucint** [reply]
@MartinZikmund, HI Martin
I was hoping that when using Uno over Skia we can make Uno free from winui. but I understand that maybe it is not necessarily your approach.

Design of my application is as follow (massively oversimplified):
We have many (hardware) devices (approx 30) that collect data and send them to the client app (uno UI).
Each device displays its data in a separated window.
Because the amount of data collected by a single device is massive (approx 32 channels * 100kHz = 3.2 million data points per second) and all of them must be drawn on the fly, I was thinking to do it in a separate thread, one thread per window/device.

I reconsider as well to draw data directly using Skia or Angle, rather than XAML and perform drawing in background thread to the "backbuffer bitmap" and when ready, flush it to the main UI thread and display it.

What do you think?

**Reply by MartinZikmund** [reply, marked as answer]
As the thread-per-window approach would require a larger rearchitecting of the application-windowing lifecycle, which is already pretty complex as we are (at least for now) keeping the backward compatibility with Uno.UWP, I think the alternative approach with a non-XAML rendering of the real-time data would be more suitable here 👍
